### ISSUE TITLE
[Console]::WriteLine raise an Exception: 'Multiple ambiguous overloads found for "WriteLine"...'

### ISSUE THREAD

**Issue by moh-hassan** [PowerShell/PowerShell#16186]
### Prerequisites

- [X] Write a descriptive title.
- [X] Make sure you are able to repro it on the [latest released version](https://github.com/PowerShell/PowerShell/releases)
- [X] Search the existing issues.
- [X] Refer to the [FAQ](https://github.com/PowerShell/PowerShell/blob/master/docs/FAQ.md).
- [X] Refer to [Differences between Windows PowerShell 5.1 and PowerShell](https://docs.microsoft.com/powershell/scripting/whats-new/differences-from-windows-powershell).

### Steps to reproduce

I tried to run the next script using WriteLine method in v7.4:
```
Function Test {
       $i = 3
       $j = 0
    [Console]::WriteLine("j={0} i={2}", $j,$i)
    }
```	 
 

### Expected behavior

```console
j=0 i=3
```

### Actual behavior

```console
Error Exception is raised:
```

### Error details

```console
MethodException:
Line |
   4 |      [Console]::WriteLine("j={0} i={2}", $j,$i)
     |      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
     | Multiple ambiguous overloads found for "WriteLine" and the argument count: "3".
```

The workaround solution to fix this issue is using splat parameters as :
```
 $args= @($j,$i)
 [Console]::WriteLine("j={0} i={1}",$args)  # working fine
```	
args in c# is params type.
```


### Environment data

```powershell
PS C:\Program Files\PowerShell\7> $PSVersionTable

Name                           Value
----                           -----
PSVersion                      7.1.4
PSEdition                      Core
GitCommitId                    7.1.4
OS                             Microsoft Windows 10.0.19042
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

**Comment by 237dmitry** [comment]
Try to use string type for formatting
```
Function Test {
       $i = 3
       $j = 0
    [Console]::WriteLine("j={0} i={1}", "$j","$i")
    }
Test
```
or
```
Function Test {
       $i = "3"
       $j = "0"
    [Console]::WriteLine("j={0} i={1}", $j,$i)
    }
Test
```

**Comment by moh-hassan** [comment]
@237dmitry, I think that the provided solution is another workaround ones  by converting data type to string. 
It's supposed that  Powershell  use .Net methods as used in c#. 
It's strange for powershell to fire this exception.
have a look to these cases:
```
1- [Console]::WriteLine("{0} {1}",10, [long]20)  #working, just cast one parameter to long
2- [Console]::WriteLine("{0} {1}",10, 20.0)  #working
3- [Console]::WriteLine("{0} {1}",10, [int]20)  #NOT WORKING !!! see working case 1
4- [Console]::WriteLine("{0} {1}",10, 20)  #NOT WORKING !!! see working case 2
```

**Comment by jborean93** [comment]
My understanding as why `[Console]::WriteLine("{0} {1}",10, 20)` doesn't work is because it cannot differentiate between these 2 choices:

```
static void WriteLine(string format, System.Object arg0, System.Object arg1)
static void WriteLine(char[] buffer, int index, int count)
```

When you cast any of the 2nd values to something that isn't an int it's able to determine the first option is correct but if you have a string (automatically casts to `char[]`) and 2 integers both overloads can technically work with PowerShell.

I'm not saying which behaviour is correct as I feel there's no real correct answer personally; people will have different opinions.

**Comment by moh-hassan** [comment]
@jborean93, I agree partially with you :), there is other one should be taken into account with `params object[]`
The signaure of  [Console.WriteLine](https://referencesource.microsoft.com/#mscorlib/system/console.cs,214d8f5050887fc0) in .Net(core) framework is:
```
//ideal selection method for all types including int and undefined numbers of parameters
public static void WriteLine(String format, params Object[] arg)    //why ignored by powershell
public static void WriteLine(String format, Object arg0, Object arg1)
  //other overload methods         
```
The next is working fine by casting int to object:
```
[Console]::WriteLine("{0} {1}",[object]10,  [object]20)   //working fine
```
So, signature with `params object[]`  may  be  the ideal for such a method.

**Comment by daxian-dbw** [comment]
From the specified arguments, there is no reason for PowerShell to think `(String format, params Object[] arg)` would be the best 
match. With `[Console]::WriteLine("{0} {1} {2} {3}",10, 20, 30, 40)`, powershell will choose `(String format, params Object[] arg)` as the best match.

The main ambiguity in the case of `[Console]::WriteLine("{0} {1}",10, 20)` is that the first argument is a string, which exactly matches `(String format, ...` but could be converted to `char[]`, while the rest 2 arguments are of `[int]` , which exactly match `int index, int count` but could be converted to `object`. There is no strong argument to say which overload is the best fit in this case -- both have partial matching parameter types -- and thus the resolution failed. I believe this is by design.

**Comment by moh-hassan** [comment]
@daxian-dbw 
Although this code is running without issues in c# (compiler/or scripter), It's supposed that PowerShell  run .Net(Core) methods in the same way as c# (Except there is a document shows the  limitation or consideration to run .Net(Core) methods  in PowerShell)

What is the official way to run such a code without issue ?

**Comment by SeeminglyScience** [comment]
The supported way for clarifying overloads is by casting.  Even if no cast actually happens, the compiler and method binder take it into account as intent when selecting the best candidate.

So for this I'd say probably:

```powershell
[Console]::WriteLine([string]'{0} {1}', 10, 30)
```

Is the easiest way to clarify intent.

> It's supposed that PowerShell run .Net(Core) methods in the same way as c#

PowerShell gives you the ability to interop with CLR methods but I'm not aware of any documentation that claims that it will be the same experience as C#.  There are many ways where it differs significantly (dynamic binding, type conversions are done for you, etc).

**Comment by moh-hassan** [comment]
@SeeminglyScience 
It seems that the direct cast of string as string is the only way to avoid the issue as you showed.
This code, although I show the intent that $format is string, is firing an Exception:
```
[string] $format='{0} {1}' 
[Console]::WriteLine($format,10,20); #Fire Exception 
```

**Comment by moh-hassan** [comment]
The next is `IL` generated from `LINQPad`, shows that `Console.WriteLine(String, Object, Object)` is selected
  
```
IL_0000	nop	
IL_0001	ldstr	"{0} {1}"
IL_0006	ldc.i4.s	0A  // 10
IL_0008	box	Int32
IL_000D	ldc.i4.s	14  // 20
IL_000F	box	Int32
IL_0014	call	Console.WriteLine (String, Object, Object)
IL_0019	nop	
IL_001A	ret	
```

**Comment by SeeminglyScience** [comment]
> @SeeminglyScience It seems that the direct cast of string as string is the only way to avoid the issue as you showed. This code, although I show the intent that $format is string, is firing an Exception:
> 
> ```
> [string] $format='{0} {1}' 
> [Console]::WriteLine($format,10,20); #Fire Exception 
> ```

Correct, it's the explicit cast *at the call site* that signals intent to the method binder.

> The next is `IL` generated from `LINQPad`, shows that `Console.WriteLine(String, Object, Object)` is selected

Yeah, C# is statically bound (overload is selected at compile time rather than runtime like PowerShell) and has much more strict type conversion rules.  PowerShell lets you implicitly convert from `string` and `char[]` where C# requires you to call `ToCharArray()`.

From the perspective of PowerShell, `string, object, object` is just as likely as `char[], int, int` due to conversion rules.  In C#, the former is the only possibility.

**Comment by moh-hassan** [comment]
Thanks @SeeminglyScience  for clarification.
As you said, explicit cast at the call site  is the solution for such issues even for string arguments.
